flang/lib/Lower/Runtime.cpp

```diff
@@ -75,45 +77,61 @@ void Fortran::lower::genStopStatement(
     const Fortran::parser::StopStmt &stmt) {
   auto &builder = converter.getFirOpBuilder();
   auto loc = converter.getCurrentLocation();
-  auto callee = genRuntimeFunction<mkRTKey(StopStatement)>(loc, builder);
-  auto calleeType = callee.getType();
   llvm::SmallVector<mlir::Value, 8> operands;
-  assert(calleeType.getNumInputs() == 3 &&
-         "expected 3 arguments in STOP runtime");
+  mlir::FuncOp callee;
+  mlir::FunctionType calleeType;
   // First operand is stop code (zero if absent)
   if (const auto &code =
           std::get<std::optional<Fortran::parser::StopCode>>(stmt.t)) {
-    auto expr = Fortran::semantics::GetExpr(*code);
-    assert(expr && "failed getting typed expression");
-    operands.push_back(fir::getBase(converter.genExprValue(*expr)));
+    auto expr = converter.genExprValue(*Fortran::semantics::GetExpr(*code));
+    LLVM_DEBUG(llvm::dbgs() << "stop expression: "; expr.dump();
+               llvm::dbgs() << '\n');
+    expr.match(
+        [&](const fir::CharBoxValue &x) {
+          callee = genRuntimeFunction<mkRTKey(StopStatementText)>(loc, builder);
+          calleeType = callee.getType();
+          // Creates a pair of operands for the CHARACTER and its LEN.
+          operands.push_back(
+              builder.createConvert(loc, calleeType.getInput(0), x.getAddr()));
+          operands.push_back(
+              builder.createConvert(loc, calleeType.getInput(1), x.getLen()));
+        },
+        [&](fir::UnboxedValue x) {
+          callee = genRuntimeFunction<mkRTKey(StopStatement)>(loc, builder);
+          calleeType = callee.getType();
+          auto cast = builder.createConvert(loc, calleeType.getInput(0), x);
+          operands.push_back(cast);
+        },
+        [&](auto) {
+          mlir::emitError(loc, "unhandled expression in STOP");
+          std::exit(1);
+        });
   } else {
+    callee = genRuntimeFunction<mkRTKey(StopStatement)>(loc, builder);
+    calleeType = callee.getType();
     operands.push_back(
         builder.createIntegerConstant(loc, calleeType.getInput(0), 0));
   }
+
   // Second operand indicates ERROR STOP
   bool isError = std::get<Fortran::parser::StopStmt::Kind>(stmt.t) ==
                  Fortran::parser::StopStmt::Kind::ErrorStop;
-  operands.push_back(
-      builder.createIntegerConstant(loc, calleeType.getInput(1), isError));
+  operands.push_back(builder.createIntegerConstant(
+      loc, calleeType.getInput(operands.size()), isError));
```

Comment:
I do not get why the position is `operands.size()` here and also for quiet argument.

Reply:
This can be the 2nd or 3rd argument. Quiet can be the 3rd or 4th argument. Rather than keep a maintain a separate variable for that, this uses the length of the operand list already constructed, which is, it turns out, what we really want anyway.

Reply:
OK, that make sense thanks.